### PR TITLE
Blocks support in motion

### PR DESCRIPTION
**Pull request by bikbah**
This PR adds blocks support in motion, which ables to select block statements in Go code: If, Func, Select, For, Range, Switch, TypeSwitch.

This commit in vim-go https://github.com/bikbah/vim-go/commit/792d66bbd3d97a42feefe99fc5f842026df2325a explains changes.

You can select content of block as `vib`, or select all block with `vab` actions.

### CODE REVIEW

**Comment by fatih** [comment]
Thanks @bikbah 

Before reviewing this, can you please explain what's the difference is with the currently `viB` or `vi{` ? What does it solve exactly? If you have gif that would be better.

**Comment by bikbah** [comment]
Sorry, all the difference is that `va{` nor `vaB` selects 'declaration' of block, but `vab` does. As if `vaf` selects declaration of functions, `vab` selects `if`, `for`, `switch` declaration line also.

**Comment by fatih** [comment]
Can you please examples? maybe some screenshots, I just want to understand what exactly those are. I have some plans for structs so I just want to be sure we're on a clear track.

**Comment by bikbah** [comment]
This is if select with `vab` 
This is if select with `va{`

File: main.go [hunk 1]
```diff
@@ -8,8 +8,8 @@ import (
 	"os"
 	"strings"
 
-	"github.com/fatih/motion/astcontext"
-	"github.com/fatih/motion/vim"
+	"github.com/bikbah/motion/astcontext"
+	"github.com/bikbah/motion/vim"
```
Comment (fatih) [inline main.go]: These should be `fatih`

File: astcontext/query.go [hunk 2]
```diff
@@ -39,6 +40,18 @@ func (p *Parser) Run(query *Query) (*Result, error) {
 	}
 
 	switch query.Mode {
+	case "benclosing":
+
```
Comment (fatih) [inline astcontext/query.go]: whitespace needts to be removed

File: astcontext/blocks.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+		if border.Type == leftBorderType {
+			if rightTypes == 0 {
+				return border.Block, nil
+			}
+			rightTypes--
+		} else {
+			rightTypes++
+		}
+	}
```
Comment (fatih) [inline astcontext/blocks.go]: Can't you just test it like this: https://github.com/fatih/motion/blob/master/astcontext/funcs.go#L271 ?

File: astcontext/blocks.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+	leftBorderType  borderType = "left"
+	rightBorderType borderType = "right"
+)
+
+// Border is helper struct, that represents Block's border
+type border struct {
+	Block *Block
+	Pos   *Position
+	Type  borderType
+}
```
Comment (fatih) [inline astcontext/blocks.go]: I'm not sure what this is? Can you give an explicit example or a picture of what you mean?

File: astcontext/blocks.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+			for _, f := range pkg.Files {
+				files = append(files, f)
+			}
+		}
+	}
+
+	var blocks []*Block
+	inspect := func(n ast.Node) bool {
+		switch x := n.(type) {
+		case *ast.ForStmt:
```
Comment (fatih) [inline astcontext/blocks.go]: Why don't you just look for `*ast.BlockStmt` ?

File: astcontext/blocks.go (hunk 3 above)
Comment (bikbah) [inline astcontext/blocks.go]: I'm not sure that it is the same case. Here we have nested blocks.

File: astcontext/blocks.go (hunk 5 above)
Comment (bikbah) [inline astcontext/blocks.go]: Cause in this case, we wouldn't be able to select block's declaration via `vab`, I mean select `if .....`, `for ....`.

Comment (fatih) [inline astcontext/blocks.go]: I think it's enough. When you think of it, the `if` line is starting at line: `lbrace.line` col: `rbrace.col`. 

So say `vab` on:

```go
if foo {
  x := 5
}
```

would select from `lbrace.line,rbrace.col` til `rbrace.line,rbrace.col`. Does it make sense for you ?

Comment (bikbah) [inline astcontext/blocks.go]: Yes, you're right. I think `*ast.BlockStmt` is enough.

**Comment by fatih** [comment]
Hi @bikbah 

I'm closing this as I don't have much time anymore to look at it. Also we talked personally about this as you know. Thanks for the initial work.

**Pull request closed without merging**